keps/sig-auth/3766-referencegrant/README.md

```diff
@@ -441,174 +368,365 @@ rejected".
+  // considered to be an additional set of objects that references can be
+  // valid to, or to put this another way, entries MUST be combined using OR.
+  //
+  // +kubebuilder:validation:MinItems=1
+  // +kubebuilder:validation:MaxItems=16
+  To []ReferenceGrantTo `json:"to"`
+
+  // Verbs describes the actions allowed by references using this grant.
+  // The following verbs are valid:
+  //
+  // * "ForwardTraffic": Allow forwarding traffic from resources described in
+  //                     "from" to resources described in "to".
+  // * "UseTLSCert":     Allow TLS Certificates found in resources described in
+  //                     "to" to be used for TLS termination by resources described
+  //                     in "from".
+  // * "PopulateData":   Allow data contained in Resources described in "to" to be
+  //                     used as a data source by resources described in "from"
+  //
+  // This resource may be used for implementation-specific use cases, in which
+  // case domain-prefixed verbs can be specified. For example (example.com/action).
+  //
+  // +kubebuilder:validation:MinItems=1
+  // +kubebuilder:validation:MaxItems=4
+  Verbs []string `json:"verbs"`
 }
 
 // +kubebuilder:object:root=true
 // ReferenceGrantList contains a list of ReferenceGrant.
 type ReferenceGrantList struct {
-	metav1.TypeMeta `json:",inline"`
-	metav1.ListMeta `json:"metadata,omitempty"`
-	Items           []ReferenceGrant `json:"items"`
+  metav1.TypeMeta `json:",inline"`
+  metav1.ListMeta `json:"metadata,omitempty"`
+  Items           []ReferenceGrant `json:"items"`
 }
 
-// ReferenceGrantSpec identifies a cross namespace relationship that is trusted
-// for Gateway API.
-type ReferenceGrantSpec struct {
-	// From describes the trusted namespaces and kinds that can reference the
-	// resources described in "To". Each entry in this list MUST be considered
-	// to be an additional place that references can be valid from, or to put
-	// this another way, entries MUST be combined using OR.
-	//
-	// +kubebuilder:validation:MinItems=1
-	// +kubebuilder:validation:MaxItems=16
-	From []ReferenceGrantFrom `json:"from"`
-
-	// To describes the resources in this namespace that may be referenced by
-	// the resources described in "From". Each entry in this list MUST be
-	// considered to be an additional set of objects that references can be
-	// valid to, or to put this another way, entries MUST be combined using OR.
-	//
-	// +kubebuilder:validation:MinItems=1
-	// +kubebuilder:validation:MaxItems=16
-	To []ReferenceGrantTo `json:"to"`
+// ReferenceGrantFrom describes trusted namespaces and resources.
+type ReferenceGrantFrom struct {
+  // +kubebuilder:validation:MinItems=1
+  // +kubebuilder:validation:MaxItems=16
+  Resources []GroupResource `json:"resources"`
+
+  // ReferenceGrantNamespace describes the namespace(s) of the referent.
+  ReferenceGrantNamespace string `json:"namespace"`
```

Comment:
Does the field name here have anything to do with the type introduced at line 444 below? What precludes this field from being named "Namespace"?